go/test/endtoend/vtgate/queries/reference/reference_test.go

```diff
@@ -156,3 +156,19 @@ func TestReferenceRouting(t *testing.T) {
 		`[[INT64(2)]]`,
 	)
 }
+
+// TestMultiReferenceQuery tests that a query with multiple references with unsharded keyspace and sharded keyspace works with join.
+func TestMultiReferenceQuery(t *testing.T) {
+	utils.SkipIfBinaryIsBelowVersion(t, 21, "vtgate")
+	conn, closer := start(t)
+	defer closer()
+
+	query :=
+		`select 1
+		 from delivery_failure df1
+         	join delivery_failure df2 on df1.id = df2.id
+         	join uks.zip_detail zd1 on df1.zip_detail_id = zd1.zip_id
+         	join uks.zip_detail zd2 on zd1.zip_id = zd2.zip_id`
```

Comment:
odd whitespaces here
```suggestion
		join delivery_failure df2 on df1.id = df2.id
		join uks.zip_detail zd1 on df1.zip_detail_id = zd1.zip_id
		join uks.zip_detail zd2 on zd1.zip_id = zd2.zip_id`
```